### PR TITLE
feat(tlc-spec-driven-tests): testing anti-cheat in the tlc-spec-driven

### PR DESCRIPTION
**Pull request by waldemarnt**
## The Problem

The `tlc-spec-driven` skill treated tests as an afterthought. Agents could generate task plans that deferred tests to the end, weaken assertions to make failing tests pass, silently delete tests, or skip them entirely — and nothing in the skill would catch it.

### Root Cause

Even after adding rules like "co-locate tests with implementation," agents kept violating them. The reason: **rules without enforcement don't work.** The rules existed in the reference files, but the task creation process (Steps 1–5) ended without ever verifying compliance. The agent followed the steps, output tasks, and never reached the validation sections at the bottom.

---

### `brownfield-mapping.md` — TESTING.md Template Enhancement
Three sections added to the template generated during brownfield mapping:
- **Test Coverage Matrix**: maps each code layer to its required test type, file location pattern, and run command. Becomes the single source of truth for what tests each task needs.
- **Parallelism Assessment**: determines which test types are safe to run concurrently across sub-agents. Prevents e2e tasks sharing a database from being marked as parallel.
- **Gate Check Commands**: tiered verification commands (Quick, Full, Build).

### `tasks.md` — Test-Aware Task Planning
The largest set of changes:
- **Step 1.5: Load Test Coverage Matrix** — inserted before task creation. Forces the agent to read TESTING.md and use the coverage matrix to assign test types to tasks. Includes the co-location rule: every task that creates a code layer with a required test type MUST include those tests in the same task.
- **Step 5: Validate Before Presenting (MANDATORY)** — the enforcement gate. Before showing tasks to the user, the agent MUST run three checks (granularity, diagram consistency, test co-location), build validation tables, and include them in the output. Any failure means restructure first — never show failing tasks.
- **Tests and Gate fields** added to the task template so every task explicitly declares its test type and gate level.
- Parallelism constraints: `[P]` flag requires the test type to be parallel-safe per the TESTING.md assessment.
- **Diagram-Definition Cross-Check**: validation table that forces the agent to prove every `Depends on` field matches the execution diagram.
- **Test Co-location Validation**: validation table that forces the agent to prove every task's `Tests` field matches the coverage matrix. Explicitly bans "tested in another task" as a justification.

### `implement.md` — RED-GREEN-VERIFY Cycle
Replaced the old "implement → self-review" flow with a strict TDD cycle:
- **Step 4 (RED)**: write test files before any implementation. Confirm they fail. If they pass before implementation exists, the tests are too weak.
- **Step 4b (GREEN)**: write minimum code to pass tests. Hard constraints: do NOT modify tests from Step 4, do NOT weaken assertions, do NOT delete or skip test cases, do NOT use the framework's skip/disable mechanism.
- **Step 5 (VERIFY)**: run the gate check. Non-zero exit code = hard stop. The test runner decides if code is correct, not the agent's self-assessment.
- **Step 6**: verify test count hasn't decreased, check for `SPEC_DEVIATION` markers.

### `validate.md` — Mandatory Test Execution
- "Run Tests (if applicable)" replaced by **"Run Full Gate Check (MANDATORY)"**. Non-zero exit code = hard stop.
- Added **Test Integrity Check**: compares test count before and after the feature. If tests decreased, investigate. Template updated with a Tests section tracking gate command, result, before/after counts, delta, and justified skips.

### `coding-principles.md` — Test Integrity Principle
New **Test Integrity** section with six explicit constraints covering the most common AI agent cheating patterns, made into explicit behavioral prohibitions. Stack-agnostic — no framework-specific syntax.

### `SKILL.md` — Context Loading + Sub-Agent Delegation
- TESTING.md added to on-demand context loading.
- New **Sub-Agent Delegation** section: documents when to delegate to sub-agents (research, implementation, parallel tasks) vs. keep in the main context (planning, validation). Specifies exactly what context each sub-agent receives and what it returns.

---

## Design Principles

1. **Data-driven, not hardcoded** — the skill detects the project's test infrastructure during brownfield mapping and uses that data to drive all decisions.
2. **Stack-agnostic** — all constraints use language-neutral phrasing ("the test framework's skip/disable/pending mechanism").
3. **The test runner decides, not the agent** — verification is deterministic (non-zero exit code = fail), not probabilistic (agent's self-assessment).
4. **Rules require enforcement** — every rule has a corresponding validation step that forces the agent to prove compliance before the user sees the output.
5. **Compilation dependencies have resolution paths** — instead of just saying "don't defer tests," the skill explains how to restructure tasks when a task genuinely can't be tested independently (merge forward or merge backward).
6. **Visibility** — validation tables are output with the tasks so the user can see the agent checked its own work.

### CODE REVIEW

File: packages/skills-catalog/skills/(development)/tlc-spec-driven/references/tasks.md [hunk 1]
```diff
@@ -190,6 +235,9 @@ T8 → T9
 - [ ] Feature works per acceptance criteria
 - [ ] Integration test passes
```
Comment (globsecure) [inline packages/skills-catalog/skills/(development)/tlc-spec-driven/references/tasks.md]: Os exemplos T2 (linha 191) e T4 (linha 236) usam o formato antigo de verificação de testes:

 - [x] Unit test passes
 - [x] Integration test passes

Mas o template canônico introduzido nesse mesmo PR (linhas 403-404) exige um formato mais estruturado:

- [x] Gate check passes: `[command from Gate Check Commands]`
- [ ] Test count: [N] tests pass (no silent deletions)

Os exemplos deviam seguir o mesmo formato do template, senão o agente vai imitar o formato antigo em vez do novo. LLMs tendem a dar mais peso a exemplos concretos do que a regras abstratas.

File: packages/skills-catalog/skills/(development)/tlc-spec-driven/references/validate.md [hunk 2]
```diff
@@ -43,11 +43,23 @@ From spec.md edge cases:
 - [ ] [Edge case 1] handled correctly
 - [ ] [Edge case 2] handled correctly
 
-### 4. Run Tests (if applicable)
+### 4. Run Full Gate Check (MANDATORY)
 
-```bash
```
Comment (globsecure) [inline packages/skills-catalog/skills/(development)/tlc-spec-driven/references/validate.md:48]: Esse passo assume que o TESTING.md já existe, mas em projetos greenfield ele não vai existir. O tasks.md (linhas 46 e 65) trata isso corretamente com condicional:

```
Read .specs/codebase/TESTING.md (if it exists)
If TESTING.md does not exist (greenfield project), ask the user what test types and commands
```

Aqui no validate.md o tom é incondicional: "Run the Build-level gate check from TESTING.md. This is NOT optional." Se o agente chegar nesse passo num projeto greenfield, não vai ter comando nenhum para rodar.

Sugestão: adicionar um fallback similar ao do tasks.md, algo como "Se TESTING.md não existe, rode os testes com o comando informado pelo usuário durante a fase de planning."

File: packages/skills-catalog/skills/(development)/tlc-spec-driven/SKILL.md [hunk 3]
```diff
@@ -112,6 +113,44 @@ Plan and implement projects with precision. Granular tasks. Clear dependencies.
...
+| Sequential tasks with no `[P]` | Yes | Keeps implementation artifacts out of the main context |
+| Planning, task creation, validation reports | No | These require the full accumulated context to be coherent |
+| Quick mode tasks | No | Too small to justify the overhead |
+
+**Context each sub-agent receives:**
+
+The orchestrating agent MUST provide each sub-agent with:
+- The specific task definition from tasks.md (What, Where, Depends on, Reuses, Done when, Tests, Gate)
+- Relevant coding principles and conventions (coding-principles.md, CONVENTIONS.md)
+- TESTING.md (for gate check commands and test patterns)
```
Comment (globsecure) [inline packages/skills-catalog/skills/(development)/tlc-spec-driven/SKILL.md]: A linha 134 diz que o agente orquestrador "MUST provide each sub-agent with" o TESTING.md, se qualificação. Mas em projetos greenfield o TESTING.md não vai existir, o próprio tasks.md (linhas 46 e 65) já reconhece isso:

```
Read .specs/codebase/TESTING.md (if it exists)
If TESTING.md does not exist (greenfield project), ask the user...
```

Sugestão: trocar para algo como - TESTING.md, if it exists (for gate check commands and test patterns) para manter consistência com o tasks.md.

File: packages/skills-catalog/skills/(development)/tlc-spec-driven/references/implement.md [hunk 4]
```diff
@@ -60,25 +66,78 @@ Approach: [brief description]
...
+1. Run the gate check: `[command from task's Gate field]`
+2. Non-zero exit code = STOP. Fix the failure. Re-run. Do not proceed until green.
+3. Confirm the test count matches expectations (no tests were silently deleted or skipped)
 
-### 6. Self-Check
+**Tiered gates (from TESTING.md Gate Check Commands):**
 
-Ask: "Would senior engineer flag this as overcomplicated?"
+| Task includes            | Gate level | What runs                |
+| ------------------------ | ---------- | ------------------------ |
```
Comment (globsecure) [inline packages/skills-catalog/skills/(development)/tlc-spec-driven/references/implement.md]: A tabela de gate levels cobre três cenários (unit, e2e/integration, last task), mas não define qual gate level usar quando o campo Tests da task é none (ex: tasks de config, entidades sem lógica). O step 4 (linha 84) corretamente manda pular para 4b nesses casos, mas o step 5 (linha 109) diz que o gate check é "MANDATORY — not 'if applicable'". O agente vai chegar no step 5 com uma task Tests: none, consultar essa tabela, e não vai encontrar correspondência.

Sugestão: adicionar uma linha na tabela para o caso Tests: none, ex:

```
 | No tests (config, entities) | Build     | Build + lint only        |
```

Ou alternativamente, deixar explícito no step 5 que tasks sem testes rodam apenas o build/lint gate.

File: packages/skills-catalog/skills/(development)/tlc-spec-driven/references/tasks.md
Comment (globsecure) [inline packages/skills-catalog/skills/(development)/tlc-spec-driven/references/tasks.md:211]: O T3 cria um componente (ZComponent.tsx) mas o "Done when" não tem nenhum critério de teste. Isso contradiz a regra de Test Co-location Validation (linha 334-347) introduzida nesse mesmo PR, que diz:

```
"Tested in another task" is NOT a valid justification for Tests: none. That is test deferral — the exact anti-pattern this validation prevents.
```

E:

```
 Tasks that fail this check MUST be fixed.
```

O T3 cria um code layer (componente) que provavelmente requer testes no coverage matrix, mas o exemplo não mostra nenhum critério de teste no "Done when" nem justifica o Tests: none. Seria bom alinhar o exemplo com a regra, adicionando um critério como - [ ] Gate check passes: [command] ou deixando explícito que é um caso Tests: none com justificativa.

**Review by globsecure — CHANGES_REQUESTED** [review]
Tenho alguns pontos que precisam de ajuste:

**Principal**: Os exemplos de tasks T2/T4 usam o formato antigo de verificação (Unit test passes, Integration test passes) enquanto o template canônico introduzido nesse mesmo PR exige o formato estruturado com Gate check passes: [command] e Test count: [N] tests pass. Agentes tendem a imitar exemplos concretos em vez de regras abstratas, então os exemplos precisam seguir o template novo.

**Consistência greenfield**: O validate.md e o SKILL.md tratam o TESTING.md como incondicional ("MUST provide", "NOT optional"), mas o tasks.md corretamente usa "(if it exists)" com fallback para greenfield. Precisa alinhar, projetos greenfield não vão ter TESTING.md.

**Gate para Tests**:none: A tabela de tiered gates no implement.md não cobre o caso Tests: none, mas o step 5 diz que o gate check é mandatório. Precisa de uma linha na tabela para esse cenário.

**T3 sem testes**: O exemplo T3 cria um componente mas não tem critério de teste no "Done when", o que contradiz a regra de Test Co-location Validation que esse PR introduz.

File: packages/skills-catalog/skills/(development)/tlc-spec-driven/references/validate.md [hunk 5]
```diff
@@ -43,11 +43,23 @@ From spec.md edge cases:
...
-# project test command
-```
+Run the Build-level gate check from TESTING.md. This is NOT optional.
+
+1. Run: `[build gate command from TESTING.md]`
+2. Non-zero exit code = STOP. Do not proceed to Code Quality Check.
+3. Record results:
```
[the remark this answers is not in the document]
Comment (waldemarnt) [reply packages/skills-catalog/skills/(development)/tlc-spec-driven/references/validate.md]: Fixed in c50cbb0: renamed Step 4 heading to 'Run Build-Level Gate Check (MANDATORY)' to align with the body text.

File: packages/skills-catalog/skills/(development)/tlc-spec-driven/references/validate.md [hunk 6]
```diff
@@ -186,9 +198,13 @@ Fix tasks follow the same format as regular tasks and can be executed with the i
...
 ## Tests
 
-- **Ran**: [test command]
-- **Result**: [X] passed, [Y] failed
-- **Failures**: [list]
+- **Gate command**: [full command]
+- **Result**: [X] passed, [Y] failed, [Z] skipped
+- **Test count before feature**: [N]
+- **Test count after feature**: [M]
+- **Delta**: [+N new tests]
```
[the remark this answers is not in the document]
Comment (waldemarnt) [reply packages/skills-catalog/skills/(development)/tlc-spec-driven/references/validate.md]: Fixed in c50cbb0: changed placeholder to `+(M - N) new tests` to avoid reusing the `N` variable already used for the before-count.

File: packages/skills-catalog/skills/(development)/tlc-spec-driven/references/tasks.md [hunk 7]
```diff
@@ -131,6 +167,9 @@ T8 → T9
 - [ ] Types exported correctly
 - [ ] No TypeScript errors
 
+**Tests**: [unit/e2e/integration/none — from coverage matrix]
+**Gate**: [quick/full — from gate check commands]
```
[the remark this answers is not in the document]
Comment (waldemarnt) [reply packages/skills-catalog/skills/(development)/tlc-spec-driven/references/tasks.md]: Fixed in c50cbb0: updated all example task templates (T1–T4) to include the `build` tier: `[quick/full/build — from gate check commands]`.

File: packages/skills-catalog/skills/(development)/tlc-spec-driven/references/tasks.md [hunk 8]
```diff
@@ -271,12 +393,15 @@ Every task MUST include:
...
+**Tests**: [unit/e2e/integration/none]
+**Gate**: [quick/full/build]
 
 **Done when:**
 
 - [ ] [Specific outcome]
 - [ ] [Specific outcome]
-- [ ] Tests pass: [command]
+- [ ] Gate check passes: `[command from Gate Check Commands]`
+- [ ] Test count: [N] tests pass (no silent deletions)
```
[the remark this answers is not in the document]
Comment (waldemarnt) [reply packages/skills-catalog/skills/(development)/tlc-spec-driven/references/tasks.md:408]: Fixed in c50cbb0: kept `Gate` as the tier label (quick/full/build) as defined in the canonical template. Updated `implement.md` Step 5 to explicitly say 'look up the command for the task's Gate level in TESTING.md's Gate Check Commands section'. The `Done when` checklist retains the concrete command via `Gate check passes: [command from Gate Check Commands]`, making the intent unambiguous without splitting the field.

File: packages/skills-catalog/skills/(development)/tlc-spec-driven/references/implement.md [hunk 9]
```diff
@@ -60,25 +66,78 @@ Approach: [brief description]
...
+
+- Tests define correct behavior independently of implementation
+- Each acceptance criterion from "Done when" maps to at least one test assertion
+- Edge cases from spec.md that apply to this task get test cases too
+
+If the task does NOT include tests (e.g., entity-only, config-only), skip to Step 4b.
+
+### 4b. Implement (GREEN)
+
+Write the minimum implementation to make all tests pass.
```
[the remark this answers is not in the document]
Comment (waldemarnt) [reply packages/skills-catalog/skills/(development)/tlc-spec-driven/references/implement.md]: Fixed in c50cbb0: rephrased Step 4b to cover both cases: 'Write the minimum implementation needed to satisfy the task's success criteria: pass all relevant tests (when present) and meet the defined verification/gate checks when there are no direct tests.'

File: packages/skills-catalog/skills/(development)/tlc-spec-driven/references/implement.md [hunk 10]
```diff
@@ -60,25 +66,78 @@ Approach: [brief description]
...
-  - No scope creep
+### 5. Gate Check (VERIFY)
 
-### 5. Verify "Done When"
+Run the gate check command from the task definition. This is MANDATORY — not "if applicable."
 
-Check all criteria before marking done.
+1. Run the gate check: `[command from task's Gate field]`
+2. Non-zero exit code = STOP. Fix the failure. Re-run. Do not proceed until green.
+3. Confirm the test count matches expectations (no tests were silently deleted or skipped)
```
[the remark this answers is not in the document]
Comment (waldemarnt) [reply packages/skills-catalog/skills/(development)/tlc-spec-driven/references/implement.md:113]: Fixed in c50cbb0: clarified Step 5 instruction to 'Look up the command for the task's Gate level (quick/full/build) in TESTING.md's Gate Check Commands section, then run it'. This makes explicit that `Gate` holds the tier, and TESTING.md is the source of the actual runnable command.

File: packages/skills-catalog/skills/(development)/tlc-spec-driven/references/tasks.md (hunk 1 above)
Comment (waldemarnt) [reply packages/skills-catalog/skills/(development)/tlc-spec-driven/references/tasks.md]: Corrigido em c50cbb0: os exemplos T2 e T4 agora usam o formato canônico: `Gate check passes: [command]` e `Test count: [N] tests pass (no silent deletions)`. Os campos `Tests` e `Gate` foram preenchidos com valores concretos (`unit`/`quick` para T2, `integration`/`full` para T4) em vez dos placeholders genéricos, para que o agente imite o padrão correto.

File: packages/skills-catalog/skills/(development)/tlc-spec-driven/references/validate.md (hunk 2 above)
Comment (waldemarnt) [reply packages/skills-catalog/skills/(development)/tlc-spec-driven/references/validate.md:48]: Corrigido em c50cbb0: adicionado fallback explícito no Step 4: 'If TESTING.md does not exist (greenfield project), use the gate command agreed upon with the user during the Tasks phase.' O comando de exemplo também foi atualizado para refletir os dois cenários.

File: packages/skills-catalog/skills/(development)/tlc-spec-driven/SKILL.md (hunk 3 above)
Comment (waldemarnt) [reply packages/skills-catalog/skills/(development)/tlc-spec-driven/SKILL.md]: Corrigido em c50cbb0: trocado para `TESTING.md, if it exists (for gate check commands and test patterns)`, alinhando com o tratamento condicional já usado no tasks.md.

File: packages/skills-catalog/skills/(development)/tlc-spec-driven/references/implement.md (hunk 4 above)
Comment (waldemarnt) [reply packages/skills-catalog/skills/(development)/tlc-spec-driven/references/implement.md]: Corrigido em c50cbb0: adicionada linha na tabela de tiered gates para o caso `Tests: none`: `| No tests (config, entities, etc) | Build | Build + lint only |`. Agentes agora têm correspondência explícita para esse cenário, sem precisar inferir.

File: packages/skills-catalog/skills/(development)/tlc-spec-driven/references/tasks.md
Comment (waldemarnt) [reply packages/skills-catalog/skills/(development)/tlc-spec-driven/references/tasks.md:211]: Corrigido em c50cbb0: o exemplo T3 agora inclui `Gate check passes` e `Test count` no 'Done when', além de `Tests: unit` e `Gate: quick`. O exemplo de componente passa a ser consistente com a regra de Test Co-location Validation introduzida neste PR.

**Review by felipfr — APPROVED** [review]
